### PR TITLE
Refactor/Cut down read only processor factories level

### PR DESCRIPTION
**Pull request by asdacap**
- Instead of `(ReadOnlyTransactionProcessorEnvFactory+)`->`(ReadOnlyTransactionProcessorEnv+)`->`(Scope+)`.
- Do `(ReadOnlyTxProcessorSource)`->`(Scope+)`.
- Separate eth simulate env from `ReadOnlyTxProcessingEnvBase`. The scope is completely different.
- `ReadOnlyTxProcessorSource` pools the created `Scope` which is shared across all user except for pre warm which need a custom `IWorldState`. 

#### What types of changes does your code introduce?

- [X] Refactoring

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

#### Notes on testing

- [X] Engine hive test
- [x] Sync
- [x] Catchup perf

### CODE REVIEW

File: src/Nethermind/Nethermind.Optimism/Rpc/RegisterOptimismRpcModules.cs [hunk 1]
```diff
@@ -113,9 +113,12 @@ protected override void RegisterTraceRpcModule(IRpcModuleProvider rpcModuleProvi
             _api.SpecProvider,
             _api.PoSSwitcher,
             _api.LogManager,
-            _api.L1CostHelper,
             _api.SpecHelper,
             new Create2DeployerContractRewriter(_api.SpecHelper, _api.SpecProvider, _api.BlockTree),
+
+            // TODO: Should this get overridden globally?
+            new OptimismReadOnlyTxProcessorSource(_api.WorldStateManager, _api.BlockTree.AsReadOnly(), _api.SpecProvider, _api.LogManager, _api.L1CostHelper, _api.SpecHelper),
+
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Optimism/Rpc/RegisterOptimismRpcModules.cs:121]: @jmederosalvarado @deffrian @flcl42 ?

File: src/Nethermind/Nethermind.Consensus/Processing/BlockCachePreWarmer.cs [hunk 2]
```diff
@@ -75,8 +77,8 @@ void WarmupWithdrawals(ParallelOptions parallelOptions, IReleaseSpec spec, Block
                 Parallel.For(0, block.Withdrawals.Length, parallelOptions,
                     i =>
                     {
-                        IReadOnlyTxProcessorSource env = _envPool.Get();
-                        using IReadOnlyTxProcessingScope scope = env.Build(stateRoot);
+                        IReadOnlyTxProcessingScope scope = _envPool.Get();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Processing/BlockCachePreWarmer.cs:80]: not disposing the scope?

File: src/Nethermind/Nethermind.Consensus/Processing/BlockCachePreWarmer.cs [hunk 3]
```diff
@@ -103,12 +105,12 @@ void WarmupTransactions(ParallelOptions parallelOptions, IReleaseSpec spec, Bloc
 
                 using ThreadExtensions.Disposable handle = Thread.CurrentThread.BoostPriority();
                 Transaction tx = block.Transactions[i];
-                IReadOnlyTxProcessorSource env = _envPool.Get();
+                IReadOnlyTxProcessingScope scope = _envPool.Get();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Processing/BlockCachePreWarmer.cs:108]: not disposing the scope?

File: src/Nethermind/Nethermind.Consensus.AuRa/InitializationSteps/StartBlockProducerAuRa.cs [hunk 4]
```diff
@@ -224,9 +224,7 @@ private BlockProducerEnv GetProducerChain(ITxSource? additionalTxSource)
         BlockProducerEnv Create()
         {
             ReadOnlyBlockTree readOnlyBlockTree = _api.BlockTree.AsReadOnly();
-
-            ReadOnlyTxProcessingEnv txProcessingEnv = _api.CreateReadOnlyTransactionProcessorSource();
-            IReadOnlyTxProcessingScope scope = txProcessingEnv.Build(Keccak.EmptyTreeHash);
+            IReadOnlyTxProcessingScope scope = _api.ReadOnlyTxProcessorSource.Build(Keccak.EmptyTreeHash);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus.AuRa/InitializationSteps/StartBlockProducerAuRa.cs:227]: not disposing the scope? Maybe add it to dispose stack?

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/IReadOnlyTxProcessorSource.cs [hunk 5]
```diff
@@ -2,11 +2,13 @@
...
 
 using Nethermind.Core.Crypto;
+using Nethermind.State;
 
 namespace Nethermind.Evm.TransactionProcessing
 {
     public interface IReadOnlyTxProcessorSource
     {
         IReadOnlyTxProcessingScope Build(Hash256 stateRoot);
+        IReadOnlyTxProcessingScope Build(IWorldState worldState);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/TransactionProcessing/IReadOnlyTxProcessorSource.cs:12]: would be good to add comments what is the difference between them

File: src/Nethermind/Nethermind.Mev/Execution/TracerFactory.cs [hunk 6]
```diff
@@ -45,10 +48,7 @@ public class TracerFactory : ITracerFactory
 
         public ITracer Create()
         {
-            ReadOnlyTxProcessingEnv txProcessingEnv = new(
-                _worldStateManager, _blockTree, _specProvider, _logManager);
-
-            IReadOnlyTxProcessingScope scope = txProcessingEnv.Build(Keccak.EmptyTreeHash);
+            IReadOnlyTxProcessingScope scope = _txProcessorSource.Build(Keccak.EmptyTreeHash);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Mev/Execution/TracerFactory.cs:51]: does tracer dispose the scope?

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceModuleFactory.cs [hunk 7]
```diff
@@ -57,8 +56,7 @@ public class TraceModuleFactory(
 
     public override ITraceRpcModule Create()
     {
-        ReadOnlyTxProcessingEnv txProcessingEnv = CreateTxProcessingEnv();
-        IReadOnlyTxProcessingScope scope = txProcessingEnv.Build(Keccak.EmptyTreeHash);
+        IReadOnlyTxProcessingScope scope = txProcessorSource.Build(Keccak.EmptyTreeHash);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceModuleFactory.cs:59]: does the scope become disposed?

File: src/Nethermind/Nethermind.JsonRpc/Modules/Proof/ProofModuleFactory.cs [hunk 8]
```diff
@@ -28,29 +28,29 @@ public class ProofModuleFactory : ModuleFactoryBase<IProofRpcModule>
...
             _blockTree = blockTree.AsReadOnly();
         }
 
         public override IProofRpcModule Create()
         {
-            ReadOnlyTxProcessingEnv txProcessingEnv = new(
-                _worldStateManager, _blockTree, _specProvider, _logManager);
-
-            IReadOnlyTxProcessingScope scope = txProcessingEnv.Build(Keccak.EmptyTreeHash);
+            IReadOnlyTxProcessingScope scope = _txProcessorSource.Build(Keccak.EmptyTreeHash);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Proof/ProofModuleFactory.cs:53]: does the scope become disposed?

File: src/Nethermind/Nethermind.JsonRpc/Modules/DebugModule/DebugModuleFactory.cs [hunk 9]
```diff
@@ -74,18 +76,13 @@ public class DebugModuleFactory : ModuleFactoryBase<IDebugRpcModule>
...
     public override IDebugRpcModule Create()
     {
-        ReadOnlyTxProcessingEnv txEnv = new(
-            _worldStateManager,
-            _blockTree,
-            _specProvider,
-            _logManager);
-
-        IReadOnlyTxProcessingScope scope = txEnv.Build(Keccak.EmptyTreeHash);
+        IReadOnlyTxProcessingScope scope = _txProcessorSource.Build(Keccak.EmptyTreeHash);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/DebugModule/DebugModuleFactory.cs:85]: Does the scope become disposed?

File: src/Nethermind/Nethermind.JsonRpc.Test/Modules/TraceRpcModuleTests.cs [hunk 10]
```diff
@@ -52,9 +53,9 @@ public async Task Build(ISpecProvider? specProvider = null, bool isAura = false)
             ReceiptsRecovery receiptsRecovery =
                 new(Blockchain.EthereumEcdsa, Blockchain.SpecProvider);
             IReceiptFinder receiptFinder = new FullInfoReceiptFinder(Blockchain.ReceiptStorage, receiptsRecovery, Blockchain.BlockFinder);
-            ReadOnlyTxProcessingEnv txProcessingEnv =
+            ReadOnlyTxProcessorSource txProcessorSource =
                 new(Blockchain.WorldStateManager, Blockchain.BlockTree.AsReadOnly(), Blockchain.SpecProvider, Blockchain.LogManager);
-            IReadOnlyTxProcessingScope scope = txProcessingEnv.Build(Keccak.EmptyTreeHash);
+            IReadOnlyTxProcessingScope scope = txProcessorSource.Build(Keccak.EmptyTreeHash);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc.Test/Modules/TraceRpcModuleTests.cs:58]: using?

**Review by LukaszRozmej — COMMENTED** [review]
Mixed feelings to be honest, it is a bit more centralized, but not sure if simpler.

@OlegJakushkin can you check if it breaks any eth_simulate hive tests?

File: src/Nethermind/Nethermind.Consensus/Processing/BlockCachePreWarmer.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockCachePreWarmer.cs:80]: No, prewarmer is handled specially, the scope are long lived here.

File: src/Nethermind/Nethermind.Consensus.AuRa/InitializationSteps/StartBlockProducerAuRa.cs (hunk 4 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Consensus.AuRa/InitializationSteps/StartBlockProducerAuRa.cs:227]: Congratulation! You've found a long lived scope. At least its explicit now.

File: src/Nethermind/Nethermind.Mev/Execution/TracerFactory.cs (hunk 6 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Mev/Execution/TracerFactory.cs:51]: I don't know actually.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceModuleFactory.cs (hunk 7 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceModuleFactory.cs:59]: Not here.

**Comment by asdacap** [comment]
It make it simpler to reason about which state something is using. It also make it clear which part of the code need to change to make things scope properly.

**Comment by LukaszRozmej** [comment]
@asdacap should we revisit this, can you update the PR and check the unresolved comments?

**Comment by asdacap** [comment]
Cant remember what it do anymore.

File: src/Nethermind/Nethermind.JsonRpc/Modules/DebugModule/DebugModuleFactory.cs (hunk 9 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.JsonRpc/Modules/DebugModule/DebugModuleFactory.cs:85]: Not here. The rpc module now become the scope.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Proof/ProofModuleFactory.cs (hunk 8 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Proof/ProofModuleFactory.cs:53]: No

**Pull request closed without merging**